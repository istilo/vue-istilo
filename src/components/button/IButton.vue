<script setup lang="ts">
import { computed } from 'vue';

type Color = 'primary' | 'secondary' | 'success' | 'warning' | 'danger';

const emit = defineEmits<{
	(e: 'click', event: MouseEvent): void;
}>();

const props = withDefaults(
	defineProps<{
		/**
		 * A color that has traffic light color base on the status
		 */
		color?: Color;

		/**
		 * A site link that will conver the button to anchor
		 */
		href?: string;

		/**
		 * Make the circle that mostly used as floating action button.
		 */
		isCircle?: boolean;

		/**
		 * A modifier that will make the background to white with a border dashed
		 */
		isDashed?: boolean;

		/**
		 * Make the button unclickable and apply idle style color.
		 */
		isDisabled?: boolean;

		/**
		 * A modifier thet will make the background to white
		 */
		isOutline?: boolean;

		/**
		 * A modifier that will make the button border sharpless
		 */
		isRounded?: boolean;

		/**
		 * Make the button padding consistent
		 */
		isSquare?: boolean;

		/**
		 * Make the button look like a plain text.
		 */
		isText?: boolean;

		/**
		 * Increase or decrease the dimension and gutter of the button.
		 */
		size?: string;

		/**
		 * Label to describe the button that will be use on wcag attributes
		 */
		label?: string;
	}>(),
	{
		color: undefined,
		href: undefined,
		isCircle: false,
		isDashed: false,
		isDisabled: false,
		isOutline: false,
		isRounded: false,
		isSquare: false,
		isText: false,
		size: undefined,
		label: undefined,
	}
);

const getTagName = computed(() => {
	return props.href ? 'a' : 'button';
});

const getClasses = computed(() => {
	return [
		'i-button',
		props.color && !props.isDashed && !props.isOutline && !props.isText
			? `i-button-${props.color}`
			: '',
		props.isText ? 'i-button-text' : '',
		props.isOutline && !props.isText ? `i-button-outline i-button-outline-${props.color}` : '',
		props.isCircle && !props.isText ? 'i-button-circle' : '',
		props.isRounded && !props.isCircle ? 'i-button-rounded' : '',
		props.isDashed ? 'i-button-dashed' : '',
		props.size ? `i-button-${props.size}` : '',
		props.isSquare ? 'i-button-square' : '',
	];
});

const handleClick = (event: MouseEvent) => {
	emit('click', event);
};
</script>

<template>
	<component
		:is="getTagName"
		:class="getClasses"
		:disabled="props.isDisabled"
		:type="!props.href ? 'button' : null"
		:href="props.href ? props.href : null"
		:aria-label="label"
		:role="props.href ? 'button' : null"
		:title="label"
		@click="handleClick">
		<slot />
	</component>
</template>
