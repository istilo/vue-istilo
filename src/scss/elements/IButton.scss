@import '@/scss/helpers';
@import '@/scss/functions';
@import '@/scss/mixins';

//  ============================================================
//  Name:               Button
//  Description:        Styles for buttons
//
//  Component:          `#{$prefix}-button-*`
//
//  Sub-components:     `#{$prefix}-button-group`
//                      `#{$prefix}-floating-button`
//
//  Modifiers:          `#{$prefix}-button-default`
//                      `#{$prefix}-button-primary`
//                      `#{$prefix}-button-secondary`
//                      `#{$prefix}-button-warning`
//                      `#{$prefix}-button-danger`
//                      `#{$prefix}-button-outline-primary`
//                      `#{$prefix}-button-outline-secondary`
//                      `#{$prefix}-button-outline-warning`
//                      `#{$prefix}-button-outline-danger`
//
//
//  States:             `active`
//  ============================================================

//  ============================================================
//  Variables
//  ============================================================

$button-line-height: control-height(medium) !default;
$button-small-line-height: control-height(small) !default;
$button-medium-line-height: control-height(medium) !default;
$button-large-line-height: control-height(medium) !default;

$button-font-size: font(size) !default;
$button-small-font-size: font-size(small) !default;
$button-medium-font-size: font(size) !default;
$button-large-font-size: font-size(medium) !default;

$button-padding-horizontal: gutter(xsmall) !default;
$button-small-padding-horizontal: gutter(xsmall) !default;

$button-medium-padding-vertical: margin(xsmall) !default;
$button-medium-padding-horizontal: gutter(small) !default;

$button-large-padding-vertical: gutter(xsmall) !default;
$button-large-padding-horizontal: gutter(large) !default;

$button-border-radius: border(radius) !default;
$button-border-width: border(width) !default;
$button-border-color: border(color) !default;

$button-disabled-background: background-color(wrapper) !default;
$button-disabled-color: color(idle) !default;
$button-disabled-border-color: $global-muted-background !default;

$button-text-line-height: font(line-height) !default;
$button-text-color: color(muted) !default;
$button-text-hover-color: color(default) !default;
$button-text-disabled-color: color(idle) !default;

$button-link-color: link(base) !default;
$button-link-hover-color: link(dark) !default;
$button-link-line-height: font(line-height) !default;
$button-link-hover-text-decoration: underline !default;
$button-link-disabled-color: color(idle) !default;

$button-icon-width: $button-line-height !default;
$button-icon-background: darken(color(inverse), 10%) !default;
$button-icon-color: color(default);

$button-icon-primary-background: lighten(color(primary), 10%) !default;
$button-icon-primary-color: color(inverse);

$button-icon-secondary-background: lighten(color(secondary), 10%) !default;
$button-icon-secondary-color: color(inverse);

$button-icon-success-background: lighten(color(success), 10%) !default;
$button-icon-success-color: color(inverse);

$button-icon-warning-background: lighten(color(warning), 10%) !default;
$button-icon-warning-color: color(inverse);

$button-icon-danger-background: lighten(color(danger), 10%) !default;
$button-icon-danger-color: color(inverse);

/* ============================================================
    Component: Button
   ============================================================ */

/*
 * 1. Remove margins in Chrome, Safari, and Opera.
 * 2. Remove borders for `button`.
 * 3. Remove border-radius in Chrome.
 * 4. Address `overflow` set to `hidden` in IE.
 * 5. Correct `font` properties and `color` not being inherited for `button`.
 * 6. Remove the inheritance of text transform in Edge, Firefox, and IE.
 * 7. Style
 * 8. `line-height` is used to create a height because it also centers the text
 *     vertically for `a` elements.
 *     Better would be to use height and flexbox to center the text vertically
 *     but flex box doesn't work in Firefox on `button` elements.
 * 9. Align text if button has a width.
 * 10. Require for `a`.
 */

.#{$prefix}button {
	/* 1 */
	margin: 0;

	/* 2 */
	border: none;

	/* 3 */
	border-radius: $button-border-radius;

	/* 4 */
	overflow: visible;

	/* 5 */
	font: inherit;
	color: inherit;

	/* 6 */
	text-transform: none;

	/* 7 */
	display: inline-block;
	box-sizing: border-box;
	padding: 0 $button-padding-horizontal;
	vertical-align: middle;
	font-size: $button-font-size;

	/* 8 */
	line-height: $button-line-height;

	/* 9 */
	text-align: center;

	/* 10 */
	text-decoration: none;
	transition: background 0.3s ease-in-out;
}

.#{$prefix}button:not(:disabled) {
	cursor: pointer;
}

.#{$prefix}button:disabled {
	cursor: not-allowed;
}

/*
 * Remove the inner border and padding in Firefox
 */
.#{$prefix}button::-moz-focus-inner {
	border: 0;
	padding: 0;
}

/* Hover */
.#{$prefix}button:hover {
	/* 8 */
	text-decoration: none;
}

/* Focus */
.#{$prefix}button:focus {
	outline: none;
}

/*
 * Disabled
 * The same for all style modifiers
 */

.#{$prefix}button-default:disabled,
.#{$prefix}button-primary:disabled,
.#{$prefix}button-secondary:disabled,
.#{$prefix}button-success:disabled,
.#{$prefix}button-warning:disabled,
.#{$prefix}button-danger:disabled {
	background-color: $button-disabled-background;
	color: $button-disabled-color;
	border: 1px solid $button-disabled-border-color;

	&:hover,
	&:focus {
		background-color: $button-disabled-background;
		color: $button-disabled-color;
		border: 1px solid $button-disabled-border-color;
	}
}

/* Style modifiers
   ============================================================ */

@include create-modifiers-with-selectors(background, button);

/* Outline Style modifiers
   ============================================================ */

@include create-outline-modifiers-with-selectors(button);

/* Size Style modifiers
   ============================================================ */

.#{$prefix}button-small {
	padding: 0 $button-small-padding-horizontal;
	line-height: $button-small-line-height;
	font-size: $button-small-font-size;
}

.#{$prefix}button-medium {
	padding: $button-medium-padding-vertical $button-medium-padding-horizontal;
	line-height: $button-medium-line-height;
	font-size: $button-medium-font-size;
}

.#{$prefix}button-large {
	padding: $button-large-padding-vertical $button-large-padding-horizontal;
	line-height: $button-large-line-height;
	font-size: $button-large-font-size;
}

/* Text Button
   ============================================================ */

.#{$prefix}button-text {
	/* 1 */
	padding: 0;
	line-height: $button-text-line-height;
	background: none;
	border: none;

	/* 2 */
	color: inherit;
}

/* Hover + Focus */
.#{$prefix}button-text:hover,
.#{$prefix}button-text:focus {
	color: inherit;
}

/* Disabled */
.#{$prefix}button-text:disabled {
	color: $button-text-disabled-color !important;
	cursor: not-allowed;
}

// Default
.#{$prefix}button-text.#{$prefix}button-text-default {
	color: color(default);
}

.#{$prefix}button-text.#{$prefix}button-text-default:hover,
.#{$prefix}button-text.#{$prefix}button-text-default:focus {
	color: darken(color(default), 10%);
	border: none;
}

// Primary
.#{$prefix}button-text.#{$prefix}button-text-primary {
	color: color(primary);
}

.#{$prefix}button-text.#{$prefix}button-text-primary:hover,
.#{$prefix}button-text.#{$prefix}button-text-primary:focus {
	color: darken(color(primary), 10%);
	border: none;
}

// Danger
.#{$prefix}button-text.#{$prefix}button-text-danger {
	color: color(danger);
}

.#{$prefix}button-text.#{$prefix}button-text-danger:hover,
.#{$prefix}button-text.#{$prefix}button-text-danger:focus {
	color: darken(color(danger), 10%);
	border: none;
}

/* Text Button Modifier
   ============================================================ */

@include create-modifiers-with-selectors(text, button);

/* Icon Button
   ============================================================ */
.#{$prefix}button-square {
	height: $button-medium-line-height;
	width: $button-medium-line-height;
	line-height: $button-medium-line-height;
	padding: 0;
}

/* Link Button
   ============================================================ */

/*
 * Link
 * 1, Reset
 * 2. Style
 */

.#{$prefix}button-link {
	/* 1 */
	padding: 0;
	line-height: $button-link-line-height;
	background: none;

	/* 2 */
	color: $button-link-color;
}

/* Hover + Focus */
.#{$prefix}button-link:hover,
.#{$prefix}button-link:focus {
	color: $button-link-hover-color;
	text-decoration: $button-link-hover-text-decoration;
}

/* Disabled */
.#{$prefix}button-link:disabled {
	color: $button-link-disabled-color;
	text-decoration: none;
}

/* Group
   ============================================================ */

/*
 * 1. Using `flex` instead of `inline-block` to prevent whitespace between
 *    child elements
 * 2. Behave like button
 * 3. Create position context
 */

.#{$prefix}button-group {
	/* 1 */
	display: inline-flex;

	/* 2 */
	vertical-align: middle;

	/* 3 */
	position: relative;
}

/* Ghost Button
   ============================================================ */

.#{$prefix}button-ghost {
	background-color: transparent;
	border: 1px dashed border(color);
	color: color(muted);
}

.#{$prefix}button-ghost:hover,
.#{$prefix}ost-button-ghost:focus {
	background-color: transparent;
	border-color: darken(border(color), 10%);
	color: color(muted);
}

/* Icon Button
   ============================================================ */

/*
 * 1. Remove the gutters
 * 2. Make the dimension equal
 * 3. Remove the default line-height
 */
.#{$prefix}button-icon {
	/* 1 */
	padding: 0;

	/* 2 */
	height: $button-line-height;
	width: $button-line-height;

	/* 3 */
	line-height: 0;
}

.#{$prefix}button-icon.#{$prefix}button-large {
	height: $button-large-line-height;
	width: $button-large-line-height;

	line-height: $button-line-height;
}
