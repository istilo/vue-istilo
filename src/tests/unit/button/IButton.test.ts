import { describe, it, expect, beforeEach, afterEach } from 'vitest';
import { shallowMount, VueWrapper, enableAutoUnmount } from '@vue/test-utils';

import IButton from '@/components/button/IButton.vue';

enableAutoUnmount(afterEach);

describe('Istilo Button', () => {
	let wrapper: VueWrapper<any>;

	beforeEach(() => {
		wrapper = shallowMount(IButton, {
			propsData: {
				label: 'Button',
			},
			slots: {
				default: 'Button',
			},
		});

		expect(wrapper.exists()).toBe(true);
	});
	afterEach(() => {
		expect(wrapper.html()).toMatchSnapshot();
	});

	it('Should set default attributes, element and class', () => {
		expect(wrapper.element.tagName).toBe('BUTTON');
		expect(wrapper.classes()).toContain('i-button');
		expect(wrapper.text()).toBe('Button');
	});

	it('Should have WCAG attributes', () => {
		expect(wrapper.attributes('type')).toBe('button');
		expect(wrapper.attributes('aria-label')).toBe('Button');
		expect(wrapper.attributes('title')).toBe('Button');
	});

	it('Should set a color modifier class name', async () => {
		await wrapper.setProps({
			color: 'primary',
		});

		expect(wrapper.classes()).toContain('i-button-primary');
	});

	it('Should be convert into anchor instead of button but has a style of button', async () => {
		await wrapper.setProps({
			href: 'https://google.com',
		});

		expect(wrapper.element.tagName).toBe('A');
		expect(wrapper.classes()).toContain('i-button');
		expect(wrapper.attributes('href')).toBe('https://google.com');
		expect(wrapper.attributes('role')).toBe('button');
		expect(wrapper.attributes('type')).toBeUndefined();
	});

	it('Should set a circle modifier class name', async () => {
		await wrapper.setProps({
			isCircle: true,
		});

		expect(wrapper.classes()).toContain('i-button-circle');
	});

	it('Should set a dashed modifier class name', async () => {
		await wrapper.setProps({
			isDashed: true,
		});

		expect(wrapper.classes()).toContain('i-button-dashed');
	});

	it('Should set a disabled modifier attributes', async () => {
		await wrapper.setProps({
			isDisabled: true,
		});

		expect(wrapper.attributes('disabled')).toBeDefined();
	});

	it('Should set a outline modifier class name', async () => {
		await wrapper.setProps({
			isOutline: true,
		});

		expect(wrapper.classes()).toContain('i-button-outline');
	});

	it('Should set a rounded modifier class name', async () => {
		await wrapper.setProps({
			isRounded: true,
		});

		expect(wrapper.classes()).toContain('i-button-rounded');
	});

	it('Should set a square modifier class name', async () => {
		await wrapper.setProps({
			isSquare: true,
		});

		expect(wrapper.classes()).toContain('i-button-square');
	});

	it('Should set a text modifier class name', async () => {
		await wrapper.setProps({
			isText: true,
		});

		expect(wrapper.classes()).toContain('i-button-text');
	});

	it('Should set a size modifier class name', async () => {
		await wrapper.setProps({
			size: 'large',
		});

		expect(wrapper.classes()).toContain('i-button-large');
	});
});
