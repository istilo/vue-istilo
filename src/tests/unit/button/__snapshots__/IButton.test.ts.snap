// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Istilo Button > Should be convert into anchor instead of button but has a style of button 1`] = `"<a class=\\"i-button\\" disabled=\\"false\\" href=\\"https://google.com\\" aria-label=\\"Button\\" role=\\"button\\" title=\\"Button\\">Button</a>"`;

exports[`Istilo Button > Should have WCAG attributes 1`] = `"<button class=\\"i-button\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a circle modifier class name 1`] = `"<button class=\\"i-button i-button-circle\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a color modifier class name 1`] = `"<button class=\\"i-button i-button-primary\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a dashed modifier class name 1`] = `"<button class=\\"i-button i-button-dashed\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a disabled modifier attributes 1`] = `"<button class=\\"i-button\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\" disabled=\\"\\">Button</button>"`;

exports[`Istilo Button > Should set a outline modifier class name 1`] = `"<button class=\\"i-button i-button-outline i-button-outline-undefined\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a rounded modifier class name 1`] = `"<button class=\\"i-button i-button-rounded\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a size modifier class name 1`] = `"<button class=\\"i-button i-button-large\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a square modifier class name 1`] = `"<button class=\\"i-button i-button-square\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set a text modifier class name 1`] = `"<button class=\\"i-button i-button-text\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;

exports[`Istilo Button > Should set default attributes, element and class 1`] = `"<button class=\\"i-button\\" type=\\"button\\" aria-label=\\"Button\\" title=\\"Button\\">Button</button>"`;
